File: api/src/main/java/org/apache/gravitino/rel/metric/SemanticModel.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.rel.metric;

import java.util.List;
import java.util.Objects;
import javax.annotation.Nullable;
import org.apache.gravitino.annotation.Unstable;

/** A structured, OSI-compatible semantic model carried by a Metric View. */
@Unstable
public final class SemanticModel {

  private final String name;
  @Nullable private final String description;
  @Nullable private final AIContext aiContext;
  private final List<Dataset> datasets;
  private final List<Relationship> relationships;
  private final List<Metric> metrics;
  private final List<CustomExtension> customExtensions;

  private SemanticModel(
      String name,
      @Nullable String description,
      @Nullable AIContext aiContext,
      List<Dataset> datasets,
      List<Relationship> relationships,
      List<Metric> metrics,
      List<CustomExtension> customExtensions) {
    this.name = name;
    this.description = description;
    this.aiContext = aiContext;
    this.datasets = datasets;
    this.relationships = relationships;
    this.metrics = metrics;
    this.customExtensions = customExtensions;
  }

  /**
   * @return A new semantic-model builder.
   */
  public static Builder builder() {
    return new Builder();
  }

  /**
   * @return The semantic-model name.
   */
  public String name() {
    return name;
  }

  /**
   * @return The semantic-model description, or {@code null}.
   */
  @Nullable
  public String description() {
    return description;
  }

  /**
   * @return The AI context, or {@code null}.
   */
  @Nullable
  public AIContext aiContext() {
    return aiContext;
  }

  /**
   * @return An immutable, non-empty list of datasets.
   */
  public List<Dataset> datasets() {
    return datasets;
  }

  /**
   * @return An immutable list of relationships.
   */
  public List<Relationship> relationships() {
    return relationships;
  }

  /**
   * @return An immutable list of metrics.
   */
  public List<Metric> metrics() {
    return metrics;
  }

  /**
   * @return An immutable list of custom extensions.
   */
  public List<CustomExtension> customExtensions() {
    return customExtensions;
  }

  @Override
  public boolean equals(Object other) {
    if (this == other) {
      return true;
    }
    if (!(other instanceof SemanticModel)) {
      return false;
    }
    SemanticModel that = (SemanticModel) other;
    return Objects.equals(name, that.name)
        && Objects.equals(description, that.description)
        && Objects.equals(aiContext, that.aiContext)
        && Objects.equals(datasets, that.datasets)
        && Objects.equals(relationships, that.relationships)
        && Objects.equals(metrics, that.metrics)
        && Objects.equals(customExtensions, that.customExtensions);
  }

  @Override
  public int hashCode() {
    return Objects.hash(
        name, description, aiContext, datasets, relationships, metrics, customExtensions);
  }

  @Override
  public String toString() {
    return "SemanticModel{"
        + "name='"
        + name
        + '\''
        + ", description='"
        + description
        + '\''
        + ", aiContext="
        + aiContext
        + ", datasets="
        + datasets
        + ", relationships="
        + relationships
        + ", metrics="
        + metrics
        + ", customExtensions="
        + customExtensions
        + '}';
  }

  /** Builder for {@link SemanticModel}. */
  public static final class Builder {
    private String name;
    @Nullable private String description;
    @Nullable private AIContext aiContext;
    private List<Dataset> datasets;
    private List<Relationship> relationships;
    private List<Metric> metrics;
    private List<CustomExtension> customExtensions;

    private Builder() {}

    /**
     * Sets the semantic-model name.
     *
     * @param name The semantic-model name.
     * @return This builder.
     */
    public Builder withName(String name) {
      this.name = name;
      return this;
    }

    /**
     * Sets the semantic-model description.
     *
     * @param description The semantic-model description.
     * @return This builder.
     */
    public Builder withDescription(@Nullable String description) {
      this.description = description;
      return this;
    }

    /**
     * Sets the AI context.
     *
     * @param aiContext The AI context.
     * @return This builder.
     */
    public Builder withAIContext(@Nullable AIContext aiContext) {
      this.aiContext = aiContext;
      return this;
    }

    /**
     * Sets the datasets.
     *
     * @param datasets The datasets.
     * @return This builder.
     */
    public Builder withDatasets(List<Dataset> datasets) {
      this.datasets = datasets;
      return this;
    }

    /**
     * Sets the relationships.
     *
     * @param relationships The relationships.
     * @return This builder.
     */
    public Builder withRelationships(List<Relationship> relationships) {
      this.relationships = relationships;
      return this;
    }

    /**
     * Sets the metrics.
     *
     * @param metrics The metrics.
     * @return This builder.
     */
    public Builder withMetrics(List<Metric> metrics) {
      this.metrics = metrics;
      return this;
    }

    /**
     * Sets the custom extensions.
     *
     * @param customExtensions The custom extensions.
     * @return This builder.
     */
    public Builder withCustomExtensions(List<CustomExtension> customExtensions) {
      this.customExtensions = customExtensions;
      return this;
    }

    /**
     * @return The semantic model.
     */
    public SemanticModel build() {
      return new SemanticModel(
          SemanticModelUtils.requireNonBlank(name, "name"),
          description,
          aiContext,
          SemanticModelUtils.requireNonEmptyList(datasets, "datasets"),
          SemanticModelUtils.immutableList(relationships, "relationships"),
          SemanticModelUtils.immutableList(metrics, "metrics"),
          SemanticModelUtils.immutableList(customExtensions, "customExtensions"));
    }
  }
}

File: web-v2/web/src/lib/utils/semanticModel.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { isEqual, isPlainObject } from 'lodash-es'

const SOURCE_NAMESPACE_LENGTH = 2
const DEFINITION_INDENT = 2

const isNonBlankString = value => typeof value === 'string' && value.trim() !== ''

const duplicatesOf = names => {
  const seen = new Set()
  const duplicates = new Set()

  names.forEach(name => {
    if (seen.has(name)) {
      duplicates.add(name)
    }
    seen.add(name)
  })

  return [...duplicates]
}

const validateSource = (dataset, errors) => {
  const { name, source } = dataset

  if (!isPlainObject(source)) {
    errors.push(`Dataset "${name}" must have a source with a namespace and a name.`)

    return
  }

  const { namespace } = source

  if (!Array.isArray(namespace) || namespace.length !== SOURCE_NAMESPACE_LENGTH || !namespace.every(isNonBlankString)) {
    errors.push(`Dataset "${name}" must have a source namespace of exactly [catalog, schema].`)
  }

  if (!isNonBlankString(source.name)) {
    errors.push(`Dataset "${name}" must have a source with a non-empty name.`)
  }
}

const validateDatasets = (datasets, errors) => {
  datasets.forEach((dataset, index) => {
    if (!isPlainObject(dataset) || !isNonBlankString(dataset.name)) {
      errors.push(`datasets[${index}] must have a non-empty name.`)

      return
    }

    validateSource(dataset, errors)

    const fields = dataset.fields
    if (fields !== undefined && !Array.isArray(fields)) {
      errors.push(`Dataset "${dataset.name}" must declare fields as an array.`)

      return
    }

    const duplicateFields = duplicatesOf((fields || []).map(field => field?.name))
    if (duplicateFields.length) {
      errors.push(`Field names must be unique within dataset "${dataset.name}": ${duplicateFields.join(', ')}.`)
    }
  })

  const duplicateDatasets = duplicatesOf(datasets.map(dataset => dataset?.name).filter(isNonBlankString))
  if (duplicateDatasets.length) {
    errors.push(`Dataset names must be unique within the semantic model: ${duplicateDatasets.join(', ')}.`)
  }
}

const validateRelationships = (relationships, datasetNames, errors) => {
  relationships.forEach((relationship, index) => {
    if (!isPlainObject(relationship) || !isNonBlankString(relationship.name)) {
      errors.push(`relationships[${index}] must have a non-empty name.`)

      return
    }

    const { name, from, to, fromColumns, toColumns } = relationship

    ;[from, to].forEach(endpoint => {
      if (!datasetNames.has(endpoint)) {
        errors.push(`Relationship "${name}" references an unknown dataset "${endpoint}".`)
      }
    })

    if (!Array.isArray(fromColumns) || !Array.isArray(toColumns) || !fromColumns.length || !toColumns.length) {
      errors.push(`Relationship "${name}" must have non-empty fromColumns and toColumns.`)

      return
    }

    if (fromColumns.length !== toColumns.length) {
      errors.push(`Relationship "${name}" must have the same number of fromColumns and toColumns.`)
    }
  })

  const duplicates = duplicatesOf(relationships.map(relationship => relationship?.name).filter(isNonBlankString))
  if (duplicates.length) {
    errors.push(`Relationship names must be unique within the semantic model: ${duplicates.join(', ')}.`)
  }
}

const validateMetrics = (metrics, errors) => {
  metrics.forEach((metric, index) => {
    if (!isPlainObject(metric) || !isNonBlankString(metric.name)) {
      errors.push(`metrics[${index}] must have a non-empty name.`)
    }
  })

  const duplicates = duplicatesOf(metrics.map(metric => metric?.name).filter(isNonBlankString))
  if (duplicates.length) {
    errors.push(`Metric names must be unique within the semantic model: ${duplicates.join(', ')}.`)
  }
}

/**
 * Checks the structural rules the server also enforces, so obvious mistakes are reported before the
 * request is sent. The server stays the authority: it additionally resolves every dataset source
 * and its columns against the catalog, which the browser cannot do.
 *
 * @param {unknown} definition The parsed semantic model definition.
 * @returns {string[]} The problems found, empty when the definition looks well formed.
 */
export const validateSemanticModelDefinition = definition => {
  const errors = []

  if (!isPlainObject(definition)) {
    return ['The definition must be a JSON object.']
  }

  const { datasets, relationships, metrics } = definition

  if (!Array.isArray(datasets) || !datasets.length) {
    return ['The definition must declare at least one dataset.']
  }

  validateDatasets(datasets, errors)

  const datasetNames = new Set(datasets.map(dataset => dataset?.name).filter(isNonBlankString))

  if (relationships !== undefined) {
    if (Array.isArray(relationships)) {
      validateRelationships(relationships, datasetNames, errors)
    } else {
      errors.push('The definition must declare relationships as an array.')
    }
  }

  if (metrics !== undefined) {
    if (Array.isArray(metrics)) {
      validateMetrics(metrics, errors)
    } else {
      errors.push('The definition must declare metrics as an array.')
    }
  }

  return errors
}

/**
 * Parses the definition authored in the dialog and validates it.
 *
 * @param {string} text The raw JSON text.
 * @returns {{ definition: object|null, errors: string[] }} The definition, or the problems found.
 */
export const parseSemanticModelDefinition = text => {
  if (!isNonBlankString(text)) {
    return { definition: null, errors: ['The definition is required.'] }
  }

  let parsed
  try {
    parsed = JSON.parse(text)
  } catch (err) {
    return { definition: null, errors: [`The definition is not valid JSON: ${err.message}`] }
  }

  const errors = validateSemanticModelDefinition(parsed)

  return errors.length ? { definition: null, errors } : { definition: parsed, errors: [] }
}

/**
 * Renders a definition for the JSON editor.
 *
 * @param {object|null|undefined} definition The definition to render.
 * @returns {string} The pretty-printed definition, or an empty string when there is none.
 */
export const formatSemanticModelDefinition = definition => {
  if (!definition) {
    return ''
  }

  return JSON.stringify(definition, null, DEFINITION_INDENT)
}

/**
 * Builds the atomic alter batch from what actually changed, so an unchanged definition is not
 * revalidated against its sources.
 *
 * @param {object} originalData The loaded semantic model.
 * @param {object} newData The edited semantic model.
 * @returns {object[]} The ordered updates, empty when nothing changed.
 */
export const genSemanticModelUpdates = (originalData, newData) => {
  const updates = []

  if (originalData.name !== newData.name) {
    updates.push({ '@type': 'rename', newName: newData.name })
  }

  const originalComment = originalData.comment || ''
  const newComment = newData.comment || ''

  if (originalComment !== newComment) {
    updates.push({ '@type': 'updateComment', newComment: newComment === '' ? null : newComment })
  }

  const originalProperties = originalData.properties || {}
  const newProperties = newData.properties || {}

  for (const key in originalProperties) {
    if (!(key in newProperties)) {
      updates.push({ '@type': 'removeProperty', property: key })
    }
  }

  for (const key in newProperties) {
    if (originalProperties[key] !== newProperties[key]) {
      updates.push({ '@type': 'setProperty', property: key, value: newProperties[key] })
    }
  }

  if (newData.definition && !isEqual(originalData.definition, newData.definition)) {
    updates.push({ '@type': 'replaceDefinition', definition: newData.definition })
  }

  return updates
}
